Add encrypted secrets

kaspth opened the pull request:
Fixes https://github.com/rails/rails/issues/25095

Adds `config/secrets.yml.enc` using Sekrets to allow storing production app secrets right in source control. When running `bin/rails secrets:edit` the file is decrypted and opened in a tmp file that then gets destroyed after the file is encrypted again and written back to its original location. 

Differs sligthly in the proposed first-call experience in that we generate a key instead of asking for it. Also supports storing the key in `config/secrets.yml.key` instead of always going the `RAILS_MASTER_KEY`.

Pending:
- [x] Tests.
- [x] Restricting `rails s` boot when no decryption key is present.
- [x] Merging encrypted into the standard secrets after decryption.

kaspth committed:
Initial encrypted secrets.

kaspth committed:
Use $EDITOR and await edits.

Adds support for atom and other editors with open commands that
return immediately.

kaspth committed:
Add USAGE instructions to secrets:edit.

kaspth committed:
Attempt to use an app_generator to build needed files.

kaspth committed:
Inline ignore method.

Couldn't be called without moving to the generator. Prefer duplication.

kaspth committed:
Ensure we setup encrypted secrets in the general config.

kaspth committed:
Run generator when executing secrets:edit.

Adds a more pleasant first boot experience where the key and file
plus ignoring them is set up if they're not present.

kaspth committed:
Revise secrets flow.

* Add support for immediately exiting editors such as atom
  (though done crudely with Time.now), while keeping vim etc. happy.
* Prefill content if the file has none (in time, I'd like to inline sekrets
  with proper attribution and making it easier to do this prefill before the editing).
* Add puts statement to guide clue the user in to what's happening.

kaspth commented on railties/lib/rails/commands/secrets/edit_command.rb:
```diff
@@ -0,0 +1,71 @@
+require "active_support/core_ext/string/strip"
+require "rails/generators/rails/app/app_generator"
+
+require "rails/engine"
```
Sekrets depends on Rails::Engine out of the box.

kaspth commented:
Later I'd like to inline just the parts of Sekrets we need, since it seems to a lot more than we require as well as pull in a number of dependencies (nothing wrong with that, but it's best Rails keeps slim).

@ahoward how would you feel about that? We'd inline the parts we're using into it's own private namespace and give you credit with: `# Adapted from Ara T. Howard's Sekrets gem.`.

Also: as a heads up, it seems Sekrets isn't distributed with a license :)

kaspth commented:
Dang, I've got trouble getting https://github.com/kaspth/rails/blob/226db6a15bc5ce5aca1854d7d2a00ab221795f86/railties/lib/rails/application.rb#L388 to recognize `config/secrets.yml.key`. `config.paths` don't seem to be too willing to cooperate, I imagine I need to add a glob to it somehow.

maclover7 commented on railties/lib/rails/commands/secrets/USAGE:
```diff
@@ -0,0 +1,13 @@
+to edit production secrets.
+
+When the temporary file is next saved the contents are encrypted and written to
+config/secrets.yml.enc while the file itself is destroyed to protect secrets
```
`secrets from leaking.`

maclover7 reviewed (changes requested):
Overall makes sense to me, nice work @kaspth. Added a few grammar comments.

Assuming we should add some documentation about this new feature?

Re `config.paths`, I think we just need to change [this line](https://github.com/rails/rails/blob/master/railties/lib/rails/application/configuration.rb#L83).

dhh commented:
I don't understand why `config/secrets.yml.enc` would be gitignored? Isn't the whole point to check the encrypted secrets into the repo, but keeping the key out of the repo? So think we should remove that file from gitignore, leaving only the key file there.

I'm also hesitant to generating this key upfront. It's pretty easy for someone then starting a new project to use this feature but forget to do something about the key file, like commit it to 1Password. If they forget that, then change computers, the encrypted data is inaccessible. IMO, better that there's a kick-off for the first time you use encrypted secrets that explains that we're generating the key, putting it there, and you should store it securely somewhere outside of the repo.

Otherwise I'm liking the simplicity of this a lot!

dhh commented:
@jeremy Could you review as well?

dhh commented:
So I'd add `bin/rails secrets:new` or something like that for the initialization. Could also just combine the two steps and have `bin/rails secrets`, which would initialize if secrets.yml.enc is missing, and otherwise go to the edit session.

kaspth commented:
> I don't understand why config/secrets.yml.enc would be gitignored?

It shouldn't! That was me being a dummy, thanks 👍

matthewd commented on Gemfile.lock:
```diff
@@ -293,6 +306,12 @@ GEM
+      fattr (>= 2.2.1)
+      highline (>= 1.6.15)
+      main (>= 5.1.1)
+      map (>= 6.3.0)
```
That's a lot of deps 😐

simi replied:
I was thinking the same.

I'm not sure if this is really worth it to integrate since you can do almost all of this as described in https://github.com/ahoward/sekrets/blob/master/README when you need this feature. I think the only difference is key generation (where Sekrets will ask you for key). That can be solved in Sekrets and I think it will be enough to update guides to recommend this approach for storing secret data within project.

But maybe I'm missing something.

kaspth replied:
Agreed: https://github.com/rails/rails/pull/28038#issuecomment-280423000 😊

kaspth committed:
Add Rails::Secrets.

Helps parsing yml files and wraps Sekrets to make it easier to remove later.

kaspth committed:
Don't ignore the encrypted secrets.

The whole point of this exercise is to add it to version control! 😇

kaspth committed:
Join Rails.root to find the right key path.

kaspth committed:
Inline Sekrets all secrets usage into Rails::Secrets.

Guess who's gonna die next!

kaspth committed:
Poof! …and then sekrets was gone.

Removes sekrets as a dependency but leaves an attribution in its place.

kaspth commented on railties/lib/rails/application/configuration.rb:
```diff
@@ -80,7 +80,7 @@ def paths
           paths = super
           paths.add "config/database",    with: "config/database.yml"
-          paths.add "config/secrets",     with: "config/secrets.yml"
+          paths.add "config/secrets",     with: "config", glob: "secrets.yml{,.enc}"
```
@maclover7 thanks for the tip!

kaspth committed:
Use a wonky way to find the right root.

`Rails` isn't guaranteed to respond to root in tests.

kaspth commented:
We could potentially try to remove the need for the `config/secrets.yml.enc` file entirely and just enable encryption on string keys in `config/secrets.yml`.

So when reading `config/secrets.yml` we traverse it and any string values starting with a signature ala [this](https://github.com/rails/rails/pull/16462/files#diff-0da514c8bf2ba2f670bddf6514a52fa7R30) we decrypt. Then on write we just encrypt every string value.

kaspth committed:
There's got to be a better way to do this 🤔

kaspth committed:
Don't use deprecated Cipher::Cipher reference.

kaspth committed:
Assign root in initializer.

That's better!

kaspth committed:
Ensure key stays within the cipher's key_len.

Ruby 2.4 enforces a specific key length. Handle this restriction by generating
correct keys from the get go.

Also skip the needless key digest (it's already hex'ed), and adopt
ActiveSupport::MessageEncryptor's aes-256-cbc encryption mode.

kaspth committed:
Test new and `app:update`ing apps encrypted secrets flow.

The config/secrets.yml, config/secrets.yml.key, config/secrets.yml.enc
files should be there.

Expect .key to contain a suitably hex'y key.

Expect .enc to be encrypted, the inspected contents are only there after
reading and *then* decrypting. Not just at read time.

Finally, expect .gitignore to not contain the generated key file.

kaspth committed:
Split secrets:edit into secrets:setup and secrets:edit.

Also add a general secrets command that does setup first, then
edits.

Required some more finagling to the command infrastructure.

kaspth commented on railties/lib/rails/commands/secrets/secrets_command.rb:
```diff
@@ -0,0 +1,56 @@
+        Rails::Generators::AppGenerator.new([ Rails.root ]).setup_encrypted_secrets
+      end
+
+      def edit
```
Neatness: this change also spills into improvements to the command infrastructure.

Previously we couldn't handle `bin/rails secrets`, `bin/rails secrets:setup` and `bin/rails secrets:edit` in one command file. But with some "smarter" lookup we can 💪

(I say smarter because the command lookups could use some simplifying but that's too much for this PR. So pour more coal into this 🚂, less changing lanes.)

kaspth committed:
Move encrypted secrets to own generator.

Allows us to much more easily explain what's happening.

Also makes this an manual opt-in for updating apps.

kaspth commented:
Split out the `secrets:edit` setup parts into a `secrets:setup` command. Here's what it looks like:

<img width="910" alt="screen shot 2017-02-19 at 20 52 44" src="https://cloud.githubusercontent.com/assets/350807/23105811/7e88c83e-f6e5-11e6-8b31-53ae7121cf11.PNG">

And if you're not using Git:

<img width="911" alt="screen shot 2017-02-19 at 20 54 25" src="https://cloud.githubusercontent.com/assets/350807/23105833/a7d08420-f6e5-11e6-840d-34b196fd10e0.PNG">

I think we could try to match other version control ignore files via `/^.\w+ignore`, but I don't know how big of deal that support would be.

kaspth commented:
The extra explanation now doesn't read too great on `rails new` though, so I wonder if we should provide a shush mode:

```ruby
      create  config/initializers/cors.rb
      create  config/initializers/filter_parameter_logging.rb
      create  config/initializers/inflections.rb
      create  config/initializers/mime_types.rb
      create  config/initializers/new_framework_defaults.rb
      create  config/initializers/wrap_parameters.rb
      create  config/locales
      create  config/locales/en.yml
Adding config/secrets.yml.key to store the encryption key: 2339a6ec44ddb46c2333b0ba3fd1c68c

Save this in a password manager your team can access.

If you lose the key, no one, including you, can access any encrypted secrets.

      create  config/secrets.yml.key

Ignoring config/secrets.yml.key so it won't end up in Git history:

      append  .gitignore

Adding config/secrets.yml.enc to store secrets that needs to be encrypted.

      create  config/secrets.yml.enc

For now the file contains this but it's been encrypted with the generated key:

production:
  secret_key_base: da84a3f26dc5ad272160fc632f83c2c976eb5de26bbe1ff7a757a5a9f9d1f1711c3eb8f2988ec841c8dc4bf45f1ef8e11c99bb3dbc3f3612a9a1abc812f306d3

You can edit encrypted secrets with `bin/rails secrets:edit`.
      create  config/boot.rb
      create  config/database.yml
      create  db
      create  db/seeds.rb
      create  lib
```

kaspth commented:
Also: made it such that updating apps have to run `rails secrets:setup` manually instead of rolling it into `rails app:update` because there's a lot of explanation that can get lost in a stream of a lot of other config updates.

dhh commented:
Just ran this on a new skeleton. I don't think we should trigger it by default. The instructions are good, but they appear inline and quickly scroll off. I think we should delay all configuration until someone needs to work with encrypted keys.

kaspth committed:
Don't start new apps with encrypted secrets.

Apps can enroll with `secrets:setup` whenever they want.

(Manually delete "encrypted_secrets" to hide the `bin/rails generate encrypted_secrets`
interface.)

kaspth committed:
Rewrite secrets help output.

kaspth committed:
Expand missing key error message.

kaspth committed:
Test that a new key isn't generated if setup is run twice.

kaspth commented:
@dhh yup, agreed. It's now purely opt-in via `bin/rails secrets:setup`.

As for the `rails server` restriction should that also happen in development or just production? Currently it happens in both if the key is missing.

kaspth committed:
[ci skip] gemspec less spacy

kaspth committed:
bring the hacksaw, woodcutter, grinder to the USAGE.

dhh commented:
Should probably make it a config setting that by default is only on in production. 

> On Feb 20, 2017, at 21:28, Kasper Timm Hansen <notifications@github.com> wrote:
> 
> @dhh yup, agreed. It's now purely opt-in via bin/rails secrets:setup.
> 
> As for the rails server restriction should that also happen in development or just production? Currently it happens in both if the key is missing.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

kaspth commented:
Yeah, I think we can borrow the value of `config.consider_all_requests_local` for this.

maclover7 commented on railties/lib/rails/secrets.rb:
```diff
@@ -0,0 +1,104 @@
+
+module Rails
+  # Greatly inspired by Ara T. Howard's magnificent sekrets gem. 😘
+  class Secrets # :nodoc:
```
Should we add some tests for `Rails::Secrets`? 😬

kaspth replied:
That's the next bit on my list, yeah. Thanks!

dhh commented:
Don't think we need to lean on that. Let's just give it its own explaining setting. 

> On Feb 20, 2017, at 21:45, Kasper Timm Hansen <notifications@github.com> wrote:
> 
> Yeah, I think we can borrow the value of config.consider_all_requests_local for this.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

kaspth committed:
Add raise_on_missing_encryption_key setting.

kaspth commented:
made it so 👉 `raise_on_missing_encryption_key`

dhh commented on railties/lib/rails/commands/secrets/USAGE:
```diff
@@ -0,0 +1,54 @@
+
+=== Editing Secrets
+
+After `bin/rails secrets:setup`, run `bin/rails secrets:edit`.
```
We should just run secrets:edit automatically after secrets:setup. There's not a time when you want to setup secrets but don't want to add any.

kaspth replied:
Opted not to do that, since an editor like `vim` takes over the screen and hides all the output where we've tried to explain what happens.

We could do `Rails::Command.invoke["secrets:edit"] if yes?("Start editing the secrets now?")` to get both.

dhh replied:
Ah, no worries. Fine to keep as-is, then.

dhh commented on railties/lib/rails/commands/secrets/USAGE:
```diff
@@ -0,0 +1,54 @@
+
+Where `config/secrets.yml.key` is fine in development, in production the ENV
+is safer. Therefore Rails also looks for the encryption key in
+`ENV["RAILS_MASTER_KEY"]`.
```
I think this is overstating things. You may well continue to use .key in production because it's being injected by Chef or Puppet or whatever. So maybe just something about how we offer the alternative.

dhh commented on railties/lib/rails/generators/rails/app/templates/config/environments/production.rb.tt:
```diff
@@ -14,6 +14,11 @@ Rails.application.configure do
+  # Aborts server start when missing the key to decrypt secrets in
+  # `config/secrets.yml.enc`. Assign the encryption key to
+  # `ENV["RAILS_MASTER_KEY"]`.
+  config.raise_on_missing_encryption_key = true
```
Let's go with `ensure_presence_of_master_key_for_encrypted_secrets`. There are many types of encryption keys. Might as well be as literal as we can.

matthewd replied:
`read_encrypted_secrets` ?

Seems needlessly confusing to silently load them in development just because the key is present

dhh replied:
I like that. Clearer and simpler.

dhh commented on railties/lib/rails/generators/rails/app/templates/config/secrets.yml:
```diff
@@ -22,9 +24,3 @@ development:
-# instead read values from the environment.
-
-production:
-  secret_key_base: <%%= ENV["SECRET_KEY_BASE"] %>
```
Should probably replace this with a pointer to encrypted secrets. Maybe something like:

```
# Do not keep production secrets in the unencrypted secrets file.
# Instead, use bin/rails secrets:setup to configure encrypted secrets.
# This is where the secret_key_base should go.
```

dhh commented:
Lovely work, this! Really simple, nice level-up that'll serve most people most of the time.

kaspth committed:
Better explain where production secrets go.

kaspth committed:
Rename config to read_encrypted_secrets.

kaspth committed:
[ci skip] more explanation.

schneems commented on railties/lib/rails/generators/rails/encrypted_secrets/templates/config/secrets.yml.enc:
```diff
@@ -0,0 +1,2 @@
+production:
+  secret_key_base: <%= SecureRandom.hex(64) %>
```
This seems problematic for containers. This value will always be different for every app instance that boots, meaning that sessions between different servers will not be able to submit forms to each other. 

Right now we default to generating secrets.yml with a `ENV['SECRET_KEY_BASE']` and it works out of the box as long as you have that env var set. Could we maybe make it so that this file is optional? I.e. if you've only got a `RAILS_MASTER_KEY` set, use that as the secret key base?

kaspth replied:
It is optional. It will only be generated when people call `bin/rails secrets:setup` manually 😊

matthewd replied:
This is a generator template; the randomness occurs once when the file's created, not each time it's read.

schneems replied:
Got it, thanks! So all I need to do is provide a `RAILS_MASTER_KEY` instead of a `RAILS_SECRET_KEY_BASE`. Can do.

kaspth committed:
Add tests for Rails::Secrets.

Go through `read_for_editing` to integration test the gauntlet of:

- reading, decrypting
- writing, encrypting, deleting intermediate decrypted file
- rereading, decrypting so everything's still legible

kaspth committed:
Stub the generated secrets.yml.enc out.

Prevents upgrading apps from breaking in production because
the production secret_key_base value was overwritten with a
new value.

claudiob commented:
Nor `rails --tasks` neither `rails --tasks -a` show the new task added by this PR.

This might make it hard for a normal user to remember "mmh… was it `rails secret`? `rails secrets`? `rails secrets:something`?"

@kaspth I can look into finding a solution for that if you think that's important.

claudiob commented on railties/lib/rails/generators/rails/encrypted_secrets/encrypted_secrets_generator.rb:
```diff
@@ -0,0 +1,66 @@
+  module Generators
+    class EncryptedSecretsGenerator < Base
+      def add_secrets_key_file
+        unless File.exist?("config/secrets.yml.key") || File.exist?("config/secrets.yml.enc")
```
@kaspth Currently, running `rails secrets:setup` after you have already run it once result in "nothing":

```sh
$ rails secrets:setup
$
```

How do you feel about raising an exception/warning if someone calls `rails secrets:setup` but the `secrets.yml.key` and/or `secrets.yml.enc` already exist?

The message could say something like "secrets already found, use `rails secrets:edit` to edit them.

kaspth replied:
If people have already opted in, then we may as well run `bin/rails secrets:edit` for them.

I'll tackle that post-merge though 👍

kaspth commented:
@claudiob don't remember us saying we support `rails --tasks`. There's a `rails help` which does show `secrets`:

```
…

All commands can be run with -h (or --help) for more information.
In addition to those commands, there are:

Rails:
  console
  dbconsole
  destroy
  generate
  new
  runner
  secrets 👈
  server
  test
  version
…
```

Then `r secrets --help` shows the documentation added in `USAGE` here. The doc handling on the new-ish command infrastructure could definitely use some sprucing up.

claudiob commented:
@kaspth Oops my bad… my head is still in the old days of `rake --tasks`. Thanks!

kaspth committed:
Test parsing skip via read_encrypted_secrets = false.

kaspth committed:
Test production reads from encrypted_secrets by default.

kaspth committed:
Fix test broken after config rename.

When renaming the secrets config to read_encrypted_secrets.

The frameworks tests break because the remove the config/environments
files, which included production.rb that had the flag enabled.

As such some of the tests couldn't find a secret_key_base.

They also couldn't a key base because the we no longer generate a secret
key base in the production environment in secrets.yml.

Fix this by assigning a key base to our config.

kaspth committed:
Assign read_encrypted_secrets up front.

Spares the mangling in the test.

kaspth committed:
Test merging encrypted secrets with precedence.

kaspth committed:
Simplify the file writes.

Skip teh blocks.

kaspth committed:
Don't disrupt the existing Heroku workflow.

People can use encrypted secrets but that's a separate action,
later.

kaspth committed:
Replace secret_key_base with a more obvious example value.

rafaelfranca reviewed (approved):
👏

simi commented:
Will be nice to add this to CHANGELOG also.

drnic commented:
Is `production.secret_key_base` value relevant if new encrypted secrets is enabled? Is it the sameor diff to `RAILS_MASTER_KEY`?

drnic commented:
Wrote a little getting-started/investigation blog post http://www.starkandwayne.com/blog/rails-5-1-applications-can-be-a-lot-more-secretive-on-cloud-foundry-and-heroku/

emboss commented on railties/lib/rails/secrets.rb:
```diff
@@ -0,0 +1,111 @@
+
+      def generate_key
+        cipher = new_cipher
+        SecureRandom.hex(cipher.key_len)[0, cipher.key_len]
```
This is limiting the key's entropy. The output of

```ruby
SecureRandom.hex(n).size # => 2n
```

is twice as long because `n` is telling `SecureRandom` "give me an output that is equivalent to `n` random bytes". Hex encoding encodes every byte as two characters, so the resulting hex string would be twice as long.

For the default AES-256 `n = 32`, and by cutting the 64 hex bytes down to 32 again, the entropy in that value is now only equivalent to 16 random bytes anymore. This makes brute forcing the key much easier: Instead of 2^256 now there are only 2^128 possible keys.

I would suggest simply using the full output of `SecureRandom.hex(cipher.key_len)` and decoding the resulting 64 hex bytes when the key is read from the file, yielding the 32-byte binary key.

nickmalcolm replied:
Looks like this has been addressed: https://github.com/rails/rails/pull/28139

emboss commented on railties/lib/rails/secrets.rb:
```diff
@@ -0,0 +1,111 @@
+        end
+
+        def new_cipher
+          OpenSSL::Cipher.new("aes-256-cbc")
```
It would be preferable to use an [AEAD](https://en.wikipedia.org/wiki/Authenticated_encryption) cipher mode like AES-GCM that also guarantees the integrity of the resulting ciphertext. As a fallback when the underlying OpenSSL version does not support GCM yet, "Encrypt-then-Mac" could be used. `ActiveSupport::MessageEncryptor` already supports this with its `#encrypt_and_sign` method - would it be possible to reuse it in this context?

Let me know what you think, I'd be happy to help.

morgoth replied:
@emboss you might be interested in https://github.com/rails/rails/pull/28139

emboss replied:
@morgoth OK, it's already being discussed, that's great, thank you for the hint!

coderanger commented:
Sorry to sealion in here but someone referred me to this PR from Twitter and it seems like you've set things up to use only a single key. This will likely make key rotations in production environments either very tricky or impossible. I would ask if you have the time before the feature goes live to make it possible to keep multiple copies of encrypted settings such that new code and new keys don't have to be updated in perfect lockstep.

Also I would recommend warning people about the dangers of storing the key in an environment variable. It does make life easier for PaaS users but caveat emptor. Tools like Airbrake and Sentry will need to be updated ASAP to not automatically log the environment variable (and configuration for each should be used in the interim). Also storing a key in an environment variable means it is passed to all subprocesses of Rails, which can be an issue for things like ImageMagick for thumbnail generation.

schneems commented:
Key rotation would be amazing. We don't have that for secret_key_base now do we?

schneems commented:
Also rollbar and sentry AFAIK use env var whitelisting so they won't need to be updated. If airbrake does something different then it has bigger problems than this PR.

coderanger commented:
Probably the best that can be done for rotation for the moment is to rework the filename structure so rather than a single `secrets.yml.enc` make room for multiple files. Once authenticated encryption of one type or another gets added, the system can then be improved to try all of them in order until one works. But having the folder structure in place now will make the transition easier even if the rotation system is added after initial release.

ahoward commented:
@coderanger - one of the main uses we've found for app specific secrets is keeping things like credit cards, private certs, etc, stored in a repo.   aka, not just *.yml based config.   this led to the development of an 'Sekrets.load(path)' api.  from there it is trivial to support Sekrets.load("#{ Rails.env }.yml") etc.   so totally agree.

morgoth commented on railties/lib/rails/secrets.rb:
```diff
@@ -0,0 +1,111 @@
+        end
+
+        def path
+          @root.join("config", "secrets.yml.enc").to_s
```
@kaspth Do you think this could be somehow configurable? I can imagine having few encrypted files based on environment

RyanSnodgrass replied:
I don't know if anyone will hear this, but I would also like to get a different master key per environment.

kaspth commented on railties/lib/rails/secrets.rb:
```diff
@@ -0,0 +1,111 @@
+        end
+
+        def path
+          @root.join("config", "secrets.yml.enc").to_s
```
I'm open to hearing that argument. But let's revisit that with a written up use case once 5.1 has been out for some time proper 😊

januszm replied:
👍  for making it configurable. Actually I would not use this feature at all until it's possible to set different encrypted secretes for production, development or staging (or any other) environment.

ce07c3 replied:
I cannot use this feature until there is a configureable option for the paths.

kaspth replied:
> I'm open to hearing that argument. But let's revisit that with a written up use case

@ce07c3 ^ all yours. Volunteer the write up if you're in a hurry 😊

ce07c3 replied:
@kaspth I have messy but working code now. I named my files `secrets.yml.enc.staging`, which clashed with `if path.end_with?(".enc")` (took a while to figure that out), but now I am wondering: are there identical by value but different sets of paths used? I did an override on `path` and `key_path` inside `Rails::Secrets class << self`, but those aren't called until later. Any idea how I am supposed to wire custom paths up? Also, how is `@read_encrypted_secrets` set?

ce07c3 replied:
@kaspth `Also, how is @read_encrypted_secrets set?` figured that out :) I was a bit tired. The rest I am still wondering about, perhaps I just need to dig a bit deeper.

morgoth replied:
@ce07c3 Are you still working on making it configurable? Can I help you somehow?

claudiob commented on railties/lib/rails/generators/rails/app/templates/config/environments/production.rb.tt:
```diff
@@ -14,6 +14,11 @@ Rails.application.configure do
+  # Attempt to read encrypted secrets from `config/secrets.yml.enc`.
+  # Requires an encryption key in `ENV["RAILS_MASTER_KEY"]` or
+  # `config/secrets.yml.key`.
+  config.read_encrypted_secrets = true
```
Hello @kaspth 😄 

The USAGE file above says:

> A `shared:` top level key is also supported such that any keys there is merged into the other environments.

This leads me to believe that any "shared" secret key will automatically be available in development.

However this is not true: encrypted secrets are currently only read in production.

**How do you feel about changing this and reading encrypted secrets in every environment?**

If you think about it, `config/secrets.yml.enc` is already organized by environment (just like config/secrets.yml), so even if you _read_ it in development, you will not be importing the keys under `production:`, only those under `shared:` and `development:`.

kaspth replied:
We skirted away from silently reading them in development just because users have a key and it's not our recommendation to use encrypted secrets as dev secrets storage.

joemsak replied:
@kaspth Could you help me understand better what you recommend for development environment? I feel like my "development" keys are still just as worthy of keeping secret and using in the same way as production, no? I tried finding something in edge guides or blog posts about upcoming rails 5.1 updates, and the PRs/issues around encrypted keys, but I don't find any advice on how to handle my test / development environments. Thanks

kaspth replied:
At the moment we don't recommend using encrypted secrets in dev/test.

joemsak replied:
Right, I got that. So, what is recommended? I need clarity on that.

Keep using Dotenv, or?

kaspth replied:
Whatever works for you 😊

januszm replied:
I'd say that the recommended way is to use the same method for storing/accessing secrets in either production or dev environment.

claudiob commented:
@kaspth @dhh 

Once I have enabled encrypted secrets, I always need to open the editor to add a new key:

```
$ rails secrets:edit
[… add a new key in the editor, save and exit …]
# New secrets encrypted and saved.
```

How do you feel about a _new feature_ to programmatically add encrypted secrets without needing an editor?

```
$ rails secrets:add API_KEY=123
# New secret API_KEY encrypted and saved.
```

The concept is very similar to the way in which [Travis encrypts environment variables](https://docs.travis-ci.com/user/environment-variables/#Encrypting-environment-variables).

dhh commented:
Claudio, hard to do right now since we have the secrets file partitioned by
shared/envs. Also, it would leave the API_KEY in your history. Which might
not be a good idea.

On Wed, Mar 22, 2017 at 2:39 PM, Claudio B. <notifications@github.com>
wrote:

> @kaspth <https://github.com/kaspth> @dhh <https://github.com/dhh>
>
> Once I have enabled encrypted secrets, I always need to open the editor to
> add a new key:
>
> $ rails secrets:edit
> [… add a new key in the editor, save and exit …]
> # New secrets encrypted and saved.
>
> How do you feel about a *new feature* to programmatically add encrypted
> secrets without needing an editor?
>
> $ rails secrets:add API_KEY=123
> # New secret API_KEY encrypted and saved.
>
> The concept is very similar to the way in which Travis encrypts
> environment variables
> <https://docs.travis-ci.com/user/environment-variables/#Encrypting-environment-variables>
> .
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rails/rails/pull/28038#issuecomment-288400498>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAAKtSSH7TTfYwzkT_I59RLTPcIKOvrAks5roSSHgaJpZM4MDZRO>
> .
>

ce07c3 commented:
I made it work. I'll see if I find the time to send a PR next week. :)

21 juni 2017 kl. 16:12 skrev Wojciech Wnętrzak <notifications@github.com>:

@morgoth commented on this pull request.

In railties/lib/rails/secrets.rb:

> +        def handle_missing_key
+          raise MissingKeyError
+        end
+
+        def read_key_file
+          if File.exist?(key_path)
+            IO.binread(key_path).strip
+          end
+        end
+
+        def key_path
+          @root.join("config", "secrets.yml.key")
+        end
+
+        def path
+          @root.join("config", "secrets.yml.enc").to_s
@ce07c3 Are you still working on making it configurable? Can I help you somehow?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.

unikitty37 commented:
Just a small point, but could somebody add a few lines to the USAGE file (and, preferably, the Rails 5.1 release notes as that's what comes up when you search the web) detailing how you actually get the secrets OUT again?

The release notes say
```
Secrets will be decrypted in production, using a key stored either in the RAILS_MASTER_KEY environment variable, or in a key file.
```
but they forget to mention where they're decrypted TO, and I suspect most people wanting to use this feature won't be too keen on having to spelunk through the 17 changed files on this PR…

ce07c3 commented:
Also, how to fix files that are broken (YAML wise but not otherwise), since exceptions are raised...

18 juli 2017 kl. 14:57 skrev John Yeates <notifications@github.com>:

Just a small point, but could somebody add a few lines to the USAGE file (and, preferably, the Rails 5.1 release notes as that's what comes up when you search the web) detailing how you actually get the secrets OUT again?

The release notes say

Secrets will be decrypted in production, using a key stored either in the RAILS_MASTER_KEY environment variable, or in a key file.
but they forget to mention where they're decrypted TO, and I suspect most people wanting to use this feature won't be too keen on having to spelunk through the 17 changed files on this PR…

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.

kaspth commented:
Please do investigate, then!

ce07c3 commented:
I did - it loads as YAML for editing purposes which is not necessary. Perhaps that's just my setup.

18 juli 2017 kl. 16:08 skrev Kasper Timm Hansen <notifications@github.com>:

Please do investigate, then!

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.

kaspth commented:
It's an encrypted yaml file. Hence it's also yaml when it's decrypted, not just for "editing purposes".

vizcay commented:
I've just updated to Rails 5.1, moved all my keys to ```config/secrets.key.enc```, updated ```.profile``` exporting ```RAILS_MASTER_KEY``` and then my deployment process started to fail because capistrano didn't have access to this variable.. I fixed it by adding ```set :default_env, { RAILS_MASTER_KEY: IO.read('config/secrets.yml.key') }``` to my ```config/deploy.rb``` file.. maybe someone finds this useful.

ce07c3 commented:
So the encryption is specific to the data format? That makes no sense to me. I should be able to read it regardless of format. After all, Rails secrets allows me to save it in an invalid format. 

> 18 juli 2017 kl. 18:26 skrev Kasper Timm Hansen <notifications@github.com>:
> 
> It's an encrypted yaml file. Hence it's also yaml when it's decrypted, not just for "editing purposes".
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>
